misc/multiaddr/src/lib.rs

```diff
@@ -150,29 +162,27 @@ impl Multiaddr {
     pub fn encapsulate<T: ToMultiaddr>(&self, input: T) -> Result<Multiaddr> {
         let new = input.to_multiaddr()?;
         let mut bytes = self.bytes.clone();
-
         bytes.extend(new.to_bytes());
-
-        Ok(Multiaddr { bytes: bytes })
+        Ok(Multiaddr { bytes })
     }
 
     /// Adds an already-parsed address component to the end of this multiaddr.
     ///
     /// # Examples
     ///
     /// ```
-    /// use multiaddr::{Multiaddr, AddrComponent};
+    /// use multiaddr::{Multiaddr, Protocol};
     ///
     /// let mut address: Multiaddr = "/ip4/127.0.0.1".parse().unwrap();
-    /// address.append(AddrComponent::TCP(10000));
+    /// address.append(Protocol::Tcp(10000));
     /// assert_eq!(address, "/ip4/127.0.0.1/tcp/10000".parse().unwrap());
     /// ```
     ///
     #[inline]
-    pub fn append(&mut self, component: AddrComponent) {
-        component.write_bytes(&mut self.bytes).expect(
-            "writing to a Vec never fails",
-        )
+    pub fn append(&mut self, p: Protocol) {
+        let mut w = Vec::new();
+        p.write_bytes(&mut w).expect("writing to a Vec never fails");
+        self.bytes.extend_from_slice(&w);
```

Comment:
`Multiaddr::append` could be further optimised for the case where `self.bytes` has an internal reference count = 1:

```rust
    pub fn append(&mut self, p: Protocol) {
        let bytes = mem::replace(&mut self.bytes, Bytes::new());
        match bytes.try_mut() {
            Ok(bytes_mut) => {
                let mut w = bytes_mut.writer();
                while let Err(e) = p.write_bytes(&mut w) {
                    assert_eq!(e.kind(), io::ErrorKind::WriteZero);
                    assert_eq!(0, w.get_ref().remaining_mut());
                    let n = w.get_ref().len();
                    w.get_mut().reserve(n * 2)
                }
                mem::replace(&mut self.bytes, w.into_inner().freeze());
            }
            Err(mut bytes) => {
                let mut w = Vec::new();
                p.write_bytes(&mut w).expect("writing to a Vec never fails");
                bytes.extend_from_slice(&w);
                mem::replace(&mut self.bytes, bytes);
            }
        }
    }
```

Since the `Write` impl for `BytesMut`  does not automatically grow its buffer, we need to handle `WriteZero` errors due to the use of `Write::write_all` in `Protocol::write_bytes`. Since `Multiaddr::append` returns unit, we have to assert that the potential `io::Error` from `Protocol::write_bytes` has `WriteZero` as its `io::ErrorKind` which is not ideal.